tanjun/parsing.py

```diff
@@ -31,8 +31,321 @@
+        if self.__options_buffer:
+            return self.__options_buffer.pop(0)
+
+        while isinstance(value := self.__seek_shlex(), str):
+            self.__arg_buffer.append(value)
+
+        return value
+
+    def __seek_shlex(self) -> typing.Union[str, typing.Tuple[str, typing.Optional[str]], None]:
+        option_name = self.__last_name
+        for value in self.__shlex:
+            is_option = value.startswith("-")
+            if is_option and option_name is not None:
+                self.__last_name = value
+                return (option_name, None)
+
+            if is_option:
+                option_name = value
+                continue
+
+            if option_name:
+                return (option_name, value)
+
+            return value
+
+        if self.__last_name is not None:
+            last_name = self.__last_name
+            self.__last_name = None
+            return (last_name, None)
+
+        return None
+
+
+async def _convert_or_default_option(
+    ctx: traits.Context, option_: traits.Parameter, value: typing.Optional[typing.Any], /
+) -> typing.Any:
+    if value is not None:
+        return await option_.convert(ctx, value)
+
+    if option_.empty_value is not undefined.UNDEFINED:
+        return option_.empty_value
+
+    raise errors.ParserError(f"Option '{option_.names[0]} cannot be empty.", option_)
+
+
+class SemanticShlex(ShlexTokenizer):
+    __slots__: typing.Sequence[str] = ("__ctx",)
+
+    def __init__(self, ctx: traits.Context, /) -> None:
+        super().__init__(ctx.content)
+        self.__ctx = ctx
+
+    async def get_arguments(self, arguments: typing.Sequence[traits.Parameter], /) -> typing.Sequence[typing.Any]:
+        results: typing.MutableSequence[typing.Any] = []
+        for argument_ in arguments:
+            results.append(await self.__process_argument(argument_))
+
+            if argument_.flags.get("greedy") or argument_.flags.get("multi"):
+                break  # Multi and Greedy parameters should always be the last parameter.
+
+        return results
+
+    async def get_options(self, options: typing.Sequence[traits.Parameter], /) -> typing.Mapping[str, typing.Any]:
+        results: typing.MutableMapping[str, typing.Any] = {}
+        raw_options = self.collect_options()
+        for option_ in options:
+            key = option_.key
+            assert key is not None  # This shouldn't ever be None for options
+            values_iter = itertools.chain.from_iterable(
+                raw_options[name] for name in option_.names if name in raw_options
+            )
+            is_multi = option_.flags.get("multi", False)
+            if is_multi and (values := list(values_iter)):
+                results[key] = asyncio.gather(
+                    *(_convert_or_default_option(self.__ctx, option_, value) for value in values)
+                )
+
+            elif not is_multi and (value := next(values_iter, None)) is not None:
```

Comment:
Maybe do another call to next(values_iter, None) and if that returns not None then raise as multiple values passed for an option which only accepts one option